### PR TITLE
feat: out of sync missing deps monitor meta

### PR DESCRIPTION
**Pull request by dkontorovskyy**
- [X] Ready for review
- [X] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [X] Reviewed by Snyk internal team

#### What does this PR do?
Preserve `missingLockFileEntry` if package.json and package-lock.json are out of sync.
The `missingDeps` are passed to registry, which knows how to display them in the project page

#### What are the relevant tickets?
[BST-780](https://snyksec.atlassian.net/browse/BST-780)

Before we didn't show anything
Screen Shot 2019-07-29 at 14 32 19

Now we show which packages are missing in package-lock.json
Screen Shot 2019-07-30 at 10 29 05

### CODE REVIEW

**Comment by lili2311** [comment]
@dkontorovskyy @orsagie please add a screenshot of Registry picking this up and rendering it now vs before?

File: src/lib/monitor.ts [hunk 1]
```diff
@@ -84,6 +84,29 @@ async function pruneTree(tree: DepTree, packageManagerName: string): Promise<Dep
   return tree;
 }
 
+function filterOutMissingDeps(depTree) {
```
Comment (lili2311) [inline src/lib/monitor.ts]: types please?

**Comment by lili2311** [comment]
Do you also show now show the same meta on the project page as for SCM?  In the project meta i the main view there should be a small alert as well

File: src/lib/monitor.ts [hunk 2]
```diff
@@ -204,8 +227,9 @@ export async function monitorGraph(
   const pluginMeta = info.plugin;
   const policyPath = meta['policy-path'] || root;
   const policyLocations = [policyPath].concat(pluckPolicies(pkg)).filter(Boolean);
+  const {filteredDepTree, missingDeps} = filterOutMissingDeps(info.package);
```
Comment (lili2311) [inline src/lib/monitor.ts]: this only applies to npm & yarn today, so let's not call this when not needed

Comment (orsagie) [reply src/lib/monitor.ts]: monitorGraph is currently only for npm. We can add another check in this method for npm || yarn package managers

File: src/lib/monitor.ts [hunk 3]
```diff
@@ -85,6 +90,29 @@ async function pruneTree(tree: DepTree, packageManagerName: string): Promise<Dep
   return tree;
 }
 
+function filterOutMissingDeps(depTree): FilteredDepTree {
```
Comment (lili2311) [inline src/lib/monitor.ts]: `filterOutMissingDeps(depTree: DepTree)` would also be nice

File: src/lib/monitor.ts [hunk 4]
```diff
@@ -201,12 +229,19 @@ export async function monitorGraph(
...
   let policy;
-  const pkg = info.package;
+  let treeMissingDeps: string[];
+  let pkg = info.package;
   const pluginMeta = info.plugin;
   const policyPath = meta['policy-path'] || root;
   const policyLocations = [policyPath].concat(pluckPolicies(pkg)).filter(Boolean);
 
-  const depGraph: depGraphLib.DepGraph = await depGraphLib.legacy.depTreeToGraph(info.package, packageManager);
+  if (meta.packageManager === 'npm') {
```
Comment (lili2311) [inline src/lib/monitor.ts]: also `yarn`?

**Pull request merged**